src/Microsoft.AspNetCore.Mvc.TagHelpers/DistributedCacheTagHelper.cs

```diff
@@ -0,0 +1,214 @@
+        public static readonly string CacheKeyPrefix = nameof(DistributedCacheTagHelper);
+        private const string NameName = "name";
+
+        /// <summary>
+        /// Creates a new <see cref="CacheTagHelper"/>.
+        /// </summary>
+        /// <param name="memoryCache">The <see cref="IMemoryCache"/>.</param>
+        /// <param name="storage">The <see cref="IDistributedCacheTagHelperStorage"/>.</param>
+        /// <param name="formatter">The <see cref="IDistributedCacheTagHelperFormatter"/>.</param>
+        /// <param name="htmlEncoder">The <see cref="HtmlEncoder"/> to use.</param>
+        public DistributedCacheTagHelper(
+            IMemoryCache memoryCache,
+            IDistributedCacheTagHelperStorage storage, 
+            IDistributedCacheTagHelperFormatter formatter,
+            HtmlEncoder htmlEncoder) 
+            : base(htmlEncoder)
+        {
+            MemoryCache = memoryCache;
+            Storage = storage;
+            Formatter = formatter;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="IDistributedCacheTagHelperStorage"/> instance used to cache entries.
+        /// </summary>
+        protected IDistributedCacheTagHelperStorage Storage { get; }
+
+        /// <summary>
+        /// Gets the <see cref="IDistributedCacheTagHelperFormatter"/> instance used to serialize the content.
+        /// </summary>
+        protected IDistributedCacheTagHelperFormatter Formatter { get; }
+
+        /// <summary>
+        /// Gets the <see cref="IMemoryCache"/> instance used to cache workers.
+        /// </summary>
+        protected IMemoryCache MemoryCache { get; }
+
+        /// <summary>
+        /// Gets or sets a unique name to discriminate cached entries.
+        /// </summary>
+        [HtmlAttributeName(NameName)]
+        public string Name { get; set; }
+        
+        /// <inheritdoc />
+        public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+
+            Task<IHtmlContent> result = null;
+            IHtmlContent content = null;
+            
+            // Create a cancellation token that will be used 
+            // to release the task from the memory cache.
+            var tokenSource = new CancellationTokenSource();
+
+            if (Enabled)
+            {
+                var key = GenerateKey(context);
+
+                byte[] value = await Storage.GetAsync(key);
+
+                if(value == null)
+                {
+                    // Is there any request already processing the value?
+                    if (!MemoryCache.TryGetValue(key, out result))
+                    {
+                        result = Task.Run(async () =>
```

Comment:
Should this use a `TaskCompletionSource`? Going by https://channel9.msdn.com/Events/TechEd/Europe/2013/DEV-B318 which recommends libraries against using `Task.Run`.